### PR TITLE
[processor/tailsampling] add missing sub policies to AND policy

### PR DESCRIPTION
**Pull request by tim-oster**
**Description:** <Describe what has changed.>
Added parsing support to AND policy for all sub policies mentioned in `AndCfg` (Latency was missing). Instead of copying the parsing code, I am suggesting a refactored approach that unifies all conversions in one place for better extensibility in the future.

Furthermore, I added error handling for both the AND & COMPOSITE policy parsing. Without it, errors during parsing were ignored and nil `PolicyEvaluator`s were appended to the sub policies which caused the process to panic once it tried to evaluate any of the two composite policies.

**Link to tracking Issue:** <Issue number if applicable>

**Testing:** <Describe what testing was performed and which tests were added.>
I ran the already existing tests to check if the config still parses correctly and fixed the COMPOSITE policy which contained invalid config.

**Documentation:** <Describe the documentation added.>

### CODE REVIEW

**Comment by pmm-sumo** [comment]
Thanks for the contribution @tim-oster! We need the CLA signed, as per [community guidelines](https://github.com/open-telemetry/community/blob/main/CONTRIBUTING.md)

**Comment by pmm-sumo** [comment]
@tim-oster could you add entry to the [changelog](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/CHANGELOG.md)?

**Comment by pmm-sumo** [comment]
This looks good overall. I like how it cleans up things

> Furthermore, I added error handling for both the AND & COMPOSITE policy parsing. Without it, errors during parsing were ignored and nil PolicyEvaluators were appended to the sub policies which caused the process to panic once it tried to evaluate any of the two composite policies.

Can you add this case to unit tests?

**Comment by jpkrohling** [comment]
I'll look at it in detail later, but I wanted to thank you for this refactoring. It was way overdue!

@mottibec, would you be interested in reviewing this one as well?

**Comment by tim-oster** [comment]
@pmm-sumo sorry for the delay! I added the two test cases and refactored the composite test so that it tests all of the implemented behavior.

File: CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,14 @@
 
 ## Unreleased
 
+### 💡 Enhancements 💡
```
Comment (jpkrohling) [inline CHANGELOG.md:5]: We now have a new way of adding changelog entries. Check the contribution guidelines, but in short, just add a new file to ./unrelesed following the pattern used by the other files in there.

File: processor/tailsamplingprocessor/and_helper.go [hunk 2]
```diff
@@ -15,46 +15,25 @@
...
 )
 
-func getNewAndPolicy(logger *zap.Logger, config AndCfg) (sampling.PolicyEvaluator, error) {
+func getNewAndPolicy(logger *zap.Logger, config *AndCfg) (sampling.PolicyEvaluator, error) {
 	var subPolicyEvaluators []sampling.PolicyEvaluator
 	for i := range config.SubPolicyCfg {
-		policyCfg := config.SubPolicyCfg[i]
-		policy, _ := getAndSubPolicyEvaluator(logger, &policyCfg)
+		policyCfg := &config.SubPolicyCfg[i]
+		policy, err := getAndSubPolicyEvaluator(logger, policyCfg)
```
Comment (jpkrohling) [inline processor/tailsamplingprocessor/and_helper.go:27]: Why not just `getSharedPolicyEvaluator(logger, &policyCfg.sharedPolicyCfg)` instead?

File: processor/tailsamplingprocessor/and_helper_test.go [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+		})
+		assert.Equal(t, expected, actual)
+	})
+
+	t.Run("unsupported sampling policy type", func(t *testing.T) {
+		_, err := getNewAndPolicy(zap.NewNop(), &AndCfg{
+			SubPolicyCfg: []AndSubPolicyCfg{
+				{
+					sharedPolicyCfg: sharedPolicyCfg{
+						Name: "test-and-policy-1",
```
Comment (jpkrohling) [inline processor/tailsamplingprocessor/and_helper_test.go:52]:
```suggestion
						Name: "test-and-policy-2",
```

File: processor/tailsamplingprocessor/composite_helper_test.go [hunk 4]
```diff
@@ -16,86 +16,71 @@ package tailsamplingprocessor
...
-						},
-						{
-							Policy:  "test-composite-policy-2",
-							Percent: 25,
-						},
+				},
+			},
+			RateAllocation: []RateAllocationCfg{
+				{
+					Policy:  "test-composite-policy-1",
```
Comment (jpkrohling) [inline processor/tailsamplingprocessor/composite_helper_test.go:49]: Is this related to the other policy-1? Or should it be policy-3?

File: processor/tailsamplingprocessor/composite_helper_test.go [hunk 5]
```diff
@@ -16,86 +16,71 @@ package tailsamplingprocessor
...
+					Percent: 25,
+				},
+				{
+					Policy:  "test-composite-policy-2",
```
Comment (jpkrohling) [inline processor/tailsamplingprocessor/composite_helper_test.go:53]: Same question here

File: CHANGELOG.md [hunk 6]
```diff
@@ -2,6 +2,14 @@
...
+
+- `tailsamplingprocessor`: Add support for all sampling policies in and & composite policy (#11505)
+
+### 🧰 Bug fixes 🧰
+
+- `tailsamplingprocessor`: Fix runtime crash caused by invalid composite policy configuration (#11505)
```
Comment (jpkrohling) [inline CHANGELOG.md:11]: I trust you fixed this, but can you point me to the test that actually asserts this?

Comment (tim-oster) [reply CHANGELOG.md]: The cause was that both the xxx_helper.go implementation did not return any errors previously. That should now be asserted by the two new `unsupported sampling policy type` test cases.

File: processor/tailsamplingprocessor/composite_helper_test.go (hunk 4 above)
Comment (tim-oster) [reply processor/tailsamplingprocessor/composite_helper_test.go:49]: I am afraid I don't understand the question. The name is equal to the name of the policy so that the rates can be properly allocated because the implementation looks up the values based on their name.

File: processor/tailsamplingprocessor/config.go [hunk 7]
```diff
@@ -68,35 +69,23 @@ type SubPolicyCfg struct {
...
+// CompositeSubPolicyCfg holds the common configuration to all policies under composite policy.
+type CompositeSubPolicyCfg struct {
+	sharedPolicyCfg `mapstructure:",squash"` // squash ensures fields are correctly decoded in embedded struct
+
+	// Configs for and policy evaluator.
+	AndCfg AndCfg `mapstructure:"and"`
+}
+
+// AndSubPolicyCfg holds the common configuration to all policies under and policy.
 type AndSubPolicyCfg struct {
```
Comment (mottibec) [inline processor/tailsamplingprocessor/config.go:87]: Do we need the `AndSubPolicyCfg` at all? if I understand correctly it's the same as `sharedPolicyCfg`

**Review by mottibec — COMMENTED** [review]
Thank you for showing the tailsampling some love!

File: processor/tailsamplingprocessor/and_helper.go (hunk 2 above)
Comment (jpkrohling) [reply processor/tailsamplingprocessor/and_helper.go:27]: Your call, but I think it's OK to have it consistent with the other similar calls. If there's a need to make a specialization of this in the future, our future selves can refactor this.

File: CHANGELOG.md (hunk 1 above)
Comment (jpkrohling) [reply CHANGELOG.md]: Looks good, just one more question: don't we have an issue tracking this? I thought I've seen an issue open for this before.

File: unreleased/tailsampling-policy-support.yaml [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+# The name of the component, or a single word describing the area of concern, (e.g. filelogreceiver)
+component: tailsampling
+
+# A brief description of the change
+note: Add support for all sampling policies in and & composite policy
+
+# One or more tracking issues related to the change
+issues: []
+
+# (Optional) One or more lines of additional information to render under the primary note.
```
Comment (jpkrohling) [inline unreleased/tailsampling-policy-support.yaml:13]: Perhaps this and the next lines could be removed?

**Comment by tim-oster** [comment]
@jpkrohling yes, absolutely. sorry for the delay.

**Comment by jpkrohling** [comment]
Ping me when it's ready for a review.

**Review by jpkrohling — APPROVED** [review]
LGTM

**Comment by tim-oster** [comment]
@jpkrohling the changelog linter complains about `FAIL: validate: specify one or more issues #'s`. I haven't found any related issues. Should I create one and add it to the changelog or this there a way to skip this step?

**Comment by jpkrohling** [comment]
@tim-oster, yes, please create one following the contribution guidelines. Basically, you'll just need to add a new file to the `unreleased` directory.

**Comment by tim-oster** [comment]
@jpkrohling sorry for the confusion. I already added the file, I meant if I should also create a new issue to track this change because otherwise I cannot link to it in the file in `/unreleased`

**Comment by jpkrohling** [comment]
You can add this PR's number. GitHub will redirect to the PR if needed.

**Review by jpkrohling — APPROVED** [review]

**Comment by mottibec** [comment]
@tim-oster @jpkrohling can we merge this? this will solve #13929
